### PR TITLE
refactor(lsp): store assets behind a mutex

### PR DESCRIPTION
**Pull request by dsherret**
This PR stores all assests behind a mutex in order to make more language server messages not require mutability.

This is possible to do because the assets are static and it is ok if two threads race to retrieve a non-cached asset from tsc.

This is slowly moving towards #11307.

### CODE REVIEW

File: cli/lsp/tsc.rs [hunk 1]
```diff
@@ -187,48 +189,115 @@ impl AssetDocument {
...
-    self.0.insert(k, v)
+    self.assets.lock().get(k).cloned()
+  }
+
+  pub(crate) async fn get(
+    &self,
+    specifier: &ModuleSpecifier,
+    // todo(dsherret): this shouldn't be a parameter, but instead retrieved via
+    // a constructor dependency
+    get_snapshot: impl Fn() -> LspResult<Arc<StateSnapshot>>,
```
Comment (dsherret) [inline cli/lsp/tsc.rs:269]: In a future PR, I'm hoping to get rid of this parameter and not have `language_server::Inner` passed around in other functions.

File: cli/lsp/tsc.rs [hunk 2]
```diff
@@ -187,48 +189,115 @@ impl AssetDocument {
   }
 }
 
+fn new_assets_map(
+) -> Arc<Mutex<HashMap<ModuleSpecifier, Option<AssetDocument>>>> {
```
Comment (bartlomieju) [inline cli/lsp/tsc.rs:193]: That's a mouthful, consider adding `type AssetsMap = HashMap<ModuleSpecifier, Option<AssetDocument>>`

**Review by kitsonk — APPROVED** [review]
LGTM

**Review by bartlomieju — APPROVED** [review]
LGTM too

**Pull request merged**